File: Android/sdk/src/main/java/com/lynx/core/base/LynxArray.java
// Copyright 2017 The Lynx Authors. All rights reserved.
package com.lynx.core.base;

import java.util.ArrayList;
import java.util.List;

public class LynxArray {
    private List<Object> mObjects;

    public LynxArray(int length) {
        mObjects = new ArrayList<>();
        for (int i = 0; i < length; i++) {
            mObjects.add(null);
        }
    }

    public LynxArray() {
        mObjects = new ArrayList<>();
    }

    public void set(int index, Object object) {
        mObjects.set(index, object);
    }

    public void add(Object object) {
        mObjects.add(object);
    }

    public int length() {
        return mObjects.size();
    }

    public Object get(int index) {
        return mObjects.get(index);
    }

    public String getComponentsType() {
        return ParamsTransform.transform(mObjects.toArray());
    }

    public void addAll(LynxArray newData){
        mObjects.addAll(newData.getList());
    }

    private List<Object> getList(){
        return mObjects;
    }

}

File: Core/runtime/base/lynx_array.cc
// Copyright 2017 The Lynx Authors. All rights reserved.

#include <sstream>
#include "lynx_array.h"

namespace jscore {
    LynxArray::LynxArray() : LynxValue(LynxValue::Type::VALUE_LYNX_ARRAY) {
        data_.lynx_array = this;
    }

    LynxArray::~LynxArray() {
        data_.lynx_array = 0;
    }

    void LynxArray::Push(LynxValue* value) {
        values_.push_back(value);
        if(value != NULL) {
            AddPtr(value);
        }
    }

    LynxValue* LynxArray::Pop() {
        if (values_.size() > 0) {
            LynxValue* value = values_[values_.size() - 1];
            values_.pop_back();
            RemovePtr(value);
            return value;
        }
        return 0;
    }

    LynxValue* LynxArray::Get(int index) {
        return values_[index];
    }

    base::ScopedPtr<LynxValue> LynxArray::Release(int index) {
        LynxValue* temp = values_[index];
        values_[index] = NULL;
        RemovePtr(temp);
        return jscore::LynxValue::MakeValueScoped(temp);
    }

    int LynxArray::Size() {
        return values_.size();
    }

    void LynxArray::RemovePtr(LynxValue* value) {
        bool exist = false;
        for (auto it = values_.begin(); it != values_.end(); ++it) {
            if (*it == value) {
                exist = true;
                break;
            }
        }
        if (!exist) {
            ptrs_.release(value);
        }
    }

    void LynxArray::AddPtr(LynxValue* value) {
        ptrs_.add(value);
    }

    std::string LynxArray::ToString() {
        std::stringstream stream;
        stream<<"[";
        for (int i = 0; i < Size(); ++i) {
            if (Get(i) != NULL) {
                stream << Get(i)->ToString();
            }
            if (i != Size() - 1) {
                stream << ", ";
            }
        }
        stream<<"]";
        return stream.str();
    }
}